rust/kernel/hwrng.rs

```diff
@@ -0,0 +1,223 @@
+    pub cleanup: bool,
+}
+
+/// A constant version where all values are to set to `false`, that is, all supported fields will
+/// be set to null pointers.
+pub const USE_NONE: ToUse = ToUse {
+    init: false,
+    cleanup: false,
+};
+
+/// Defines the [`Operations::TO_USE`] field based on a list of fields to be populated.
+#[macro_export]
+macro_rules! declare_hwrng_operations {
+    () => {
+        const TO_USE: $crate::hwrng::ToUse = $crate::hwrng::USE_NONE;
+    };
+    ($($i:ident),+) => {
+        const TO_USE: kernel::hwrng::ToUse =
+            $crate::hwrng::ToUse {
+                $($i: true),+ ,
+                ..$crate::hwrng::USE_NONE
+            };
+    };
+}
+
+unsafe extern "C" fn init_callback<T: Operations>(rng: *mut bindings::hwrng) -> c_types::c_int {
+    from_kernel_result! {
+        // SAFETY: `priv` private data field was initialized during creation of
+        // the `bindings::hwrng` in `Self::init_hwrng` method. This callback is only
+        // called once the driver is registered.
+        let data = unsafe { T::Data::borrow((*rng).priv_ as *const core::ffi::c_void) };
+        T::init(data)?;
+        Ok(0)
+    }
+}
+
+unsafe extern "C" fn cleanup_callback<T: Operations>(rng: *mut bindings::hwrng) {
+    // SAFETY: `priv` private data field was initialized during creation of
+    // the `bindings::hwrng` in `Self::init_hwrng` method. This callback is only
+    // called once the driver is registered.
+    let data = unsafe { T::Data::borrow((*rng).priv_ as *const core::ffi::c_void) };
+    T::cleanup(data);
+}
+
+unsafe extern "C" fn read_callback<T: Operations>(
+    rng: *mut bindings::hwrng,
+    data: *mut c_types::c_void,
+    max: usize,
+    wait: bindings::bool_,
+) -> c_types::c_int {
+    from_kernel_result! {
+        // SAFETY: `priv` private data field was initialized during creation of
+        // the `bindings::hwrng` in `Self::init_hwrng` method. This callback is only
+        // called once the driver is registered.
+        let drv_data = unsafe { T::Data::borrow((*rng).priv_ as *const core::ffi::c_void) };
+
+        // SAFETY: Slice is created from `data` and `max` arguments that are C's buffer
+        // along with its size in bytes that are safe for this conversion.
+        let buffer = unsafe { from_raw_parts_mut(data as *mut c_types::c_char, max) };
+        let ret = T::read(drv_data, buffer, wait)?;
+        Ok(ret as _)
+    }
+}
+
+// SAFETY: `Registration` does not expose any of its state across threads.
+unsafe impl<T: Operations> Sync for Registration<T> {}
+
+impl<T: Operations> Drop for Registration<T> {
+    /// Removes the registration from the kernel if it has completed successfully before.
+    fn drop(&mut self) {
+        // SAFETY: The instance of Registration<T> is unregistered only
+        // after being initialized and registered before.
+        if self.registered {
+            unsafe { bindings::hwrng_unregister(self.hwrng.get()) };
```

Comment:
After unregistration, we also need to cleanup the memory allocated for the private data (`T::Data`).

We can either do it here, or make `T::cleanup` take a `T::Data` as argument (i.e., pass ownership during `cleanup`, which would allow implementers to defer the destruction of the data). If you choose the latter, then I think `T::Data` must also implement `Send` as cleanup apparently can be called from any thread.

Reply:
Ack, thanks.